File: ObjC/notify-demo-service-extension-objc/NotificationService.m
//
//  NotificationService.m
//  notify-demo-service-extension-objc
//
//  Copyright © 2023 Mail.Ru LLC. All rights reserved.
//

#import "NotificationService.h"

@import CXHubCore;
@import CXHubNotify;

static BOOL ApiIsInitialized = true;

@implementation NotificationService

+(void)initialize {
    //Initialize CXHubSDK
    //CXAppConfig *appConfig = [CXAppConfig defaultConfig];
    //ApiIsInitialized = [CXApp initExtensionWithConfig:appConfig withEventsReceiver:nil];
    
    NSString *configFile = [[NSBundle mainBundle] pathForResource:@"Notify" ofType:@"plist"];
    CXAppConfig *config = [[CXAppConfig alloc] initWithConfig:configFile];
    ApiIsInitialized = [CXApp initExtensionWithConfig:config withEventsReceiver:nil];
}

- (void)didReceiveNotificationRequest:(UNNotificationRequest *)request
                   withContentHandler:(void (^)(UNNotificationContent * _Nonnull))contentHandler  API_AVAILABLE(ios(10.0)){
    
    if (!ApiIsInitialized) {
        //CXHubSDK wasn't initialized or some error occured
        contentHandler(request.content);
        return;
    }
    
    //If next is true, then it's CXHub notification, so library will proceed further processing
    BOOL isCXHubNotification = [CXApp didReceiveExtensionNotificationRequest:request withContentHandler:contentHandler];
    
    if(!isCXHubNotification) {
        //Some other application notification, so application could do whatever necessary
        contentHandler(request.content);
    }
}

- (void)serviceExtensionTimeWillExpire {
    //Check TTL of notification
    BOOL isTimeExpire = [CXApp serviceExtensionTimeWillExpire];
    if (isTimeExpire) {
        //In case CXHubSDK extension api logic was started CXHubSDK will take care of content handlers,
        //otherwise application could do whatever necessary
    } else {
        //Some other custom application logic to call content handler
    }
}

@end

File: Swift/notify-demo-service-extension-swift/NotificationService.swift
//
//  NotificationService.swift
//  notify-demo-sevice-extension-swift
//
//  Copyright © 2023 Mail.Ru LLC. All rights reserved.
//

import UserNotifications
import CXHubCore
import CXHubNotify

class NotificationService: UNNotificationServiceExtension {

    private var apiIsInitialized :  Bool = false
    
    override init() {
        //Init sdk
        guard let configFile = Bundle.main.path(forResource: "Notify", ofType: "plist") else { fatalError() };
        guard let config = CXAppConfig(config: configFile) else { fatalError() }
        
        //guard let config = CXAppConfig.default() else { fatalError() }
        apiIsInitialized = CXApp.initExtension(with: config, withEventsReceiver: nil)
    }
    
    override func didReceive(_ request: UNNotificationRequest, withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void) {
        
        if self.apiIsInitialized {
            //This is notification for CXHubSDK and it will handle all other logic
            let didReceive = CXApp.didReceiveExtensionNotificationRequest(request, withContentHandler: contentHandler)
            NSLog("Did receive: %@", didReceive ? "YES":"NO")
            if !didReceive {
                contentHandler(request.content)
            }
        } else {
            //Some other application notification, so application could do whatever necessary
            contentHandler(request.content)
        }
    }
    
    override func serviceExtensionTimeWillExpire() {
        if self.apiIsInitialized  {
            let timeWillExpire = CXApp.serviceExtensionTimeWillExpire()
            NSLog("Time expire: %@", timeWillExpire ? "YES":"NO")
            //In case CXHubSDK extension api logic was started, then CXHubSDK will take care of content handlers,
            //otherwise application could do whatever necessary
        } else {
            //Some other custom application logic to call content handler
        }
    }
}

